### ISSUE TITLE
Race condition with nesting

### ISSUE THREAD

**Issue by allensquare** [Accudio/async-alpine#39]
Hello,

I'm experiencing a race condition issue with nested async alpine when processing mutations via the `_mutations` setup [here](https://github.com/Accudio/async-alpine/blob/7fa2b132edad71cd98f1d52bc10b0bef91a34600/src/core/async-alpine.js#L235-L241). The problem surfaces when I try to reference a variable from the parent component in the child component.

Usually, the parent node is able to download the javascript and execute the `x-data` directive first. However, sometimes the child node executes first, leading to an error when attempting to access an unset variable. When I pause at the time of error with debugger and inspect the DOM, I notice that the child node has already had its `x-ignore` attribute removed, while the parent node still has its `x-ignore` attribute. If I continue with debugger, I can see that parent node then gets its `x-ignore` removed and `x-data` processed after.

I suspect that issue may be caused by `_componentStrategy` being async but not waited on, and so it can't guarantee that the parent node downloads the javascript and activates itself before its child components. Normally, `x-ignore` on any parent node should prevent any alpine directives from running in any child node, but because async alpine runs `Alpine.initTree` directly on the child node, it appears to ignore this.

Quick example of a setup, and let's imagine that `component2.js` references a variable that is set up in `component1.js`, such as `this.component1property`.
```
<div
    x-ignore
    ax-load
    ax-load-src="/assets/component1.js"
    x-data="component1"
>
     <div
        x-ignore
        ax-load
        ax-load-src="/assets/component2.js"
        x-data="component2"
    >
        ...
    </div>
</div>
```

I do see some historical implementation of some parent strategy that appears to deal with similar concerns, but looks like that has long been removed. Any thoughts?

**Comment by iniznet** [comment]
I can't reproduce your issue, it loads in order just fine in my simple parent-child component test with the default (eager) strategy.

You can employ the event strategy to ensure that the parent is always loaded first and then trigger an event that the child component listens to.

```html
<div
    x-ignore
    ax-load
    ax-load-src="/assets/component1.js"
    x-data="component1"
    x-init="$dispatch('async-alpine:load', { id: 'component-2' })"
>
     <div
        x-ignore
        ax-load
        ax-load-src="/assets/component2.js"
        x-data="component2"
        id="component-2"
    >
        ...
    </div>
</div>
```

Or, using a custom event, with the parent: `$dispatch('custom-event')` and child: `ax-load="event (custom-event)".

**Comment by allensquare** [comment]
@iniznet 

Here's a quick example to replicate the bug: [codesandbox](https://codesandbox.io/p/sandbox/async-alpine-inline-forked-v4zn8r?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cls1l866m00063b6fuxeta542%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cls1l866m00023b6f3cai1c07%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cls1l866m00033b6f8vpi9vjq%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cls1l866m00053b6f1xz9hr1w%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cls1l866m00023b6f3cai1c07%2522%253A%257B%2522id%2522%253A%2522cls1l866m00023b6f3cai1c07%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522cls1l866m00053b6f1xz9hr1w%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cls1l866m00043b6f2ad1i2ac%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cls1l866m00053b6f1xz9hr1w%2522%252C%2522activeTabId%2522%253A%2522cls1l866m00043b6f2ad1i2ac%2522%257D%252C%2522cls1l866m00033b6f8vpi9vjq%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cls1l866m00033b6f8vpi9vjq%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Try refreshing the page multiple times and watching the console errors. On good loads with no errors, clicking the increment button will increase both numbers. However, if you see the console error `Uncaught ReferenceError: component1property is not defined` and try to click the increment button, you'll see that only the first one increments.

**Comment by allensquare** [comment]
Basically the wrong `_x_dataStack` is being built during [`addScopeToNode`](https://github.com/alpinejs/alpine/blob/2f59ee2bc115b32bee8cdd67c6edd6f31d01214c/packages/alpinejs/src/scope.js#L6) for the child node. If you inspect the child node's properties, you'll see that it's missing sometimes.

**Comment by Accudio** [comment]
Hmmm good catch @allensquare.

You're correct in that a previous Async Alpine iteration could load components independently of their parent elements, by using a strategy of renaming Alpine attributes. I removed that functionality in version 2.0 in favour of relying on `x-ignore` for a simpler and faster implementation.

I think it's important to retain the asynchronous behaviour of component downloads to prevent waterfalls, but we definitely need to come up with a solution to only allow child components to run when they're able to.

Without digging into the code — I can't at the moment — I think the solution here is probably that we shouldn't run `initTree` if the current component has `x-ignore` in its ancestors.
I think we should be able to register the component and remove `x-ignore` safely since the parent is still preventing Alpine from running it. Then when the parent component is initialised, it will run both itself and the child component.

I can probably look into this in more depth later this week, or could review a PR.

It is curious if `initTree` allows us to run a child component even with an `x-ignore`d ancestor—it doesn't seem like intended functionality but I wonder if that would allow a reliable way to re-add the ability to load wrapper components independently of child components. With version 1.0 I did use that where different JS was required for different screen sizes but they could use the same markup. That's a separate conversation and experiment however!
